keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+1. creation of a ConfigMap, residing in the OBC's namespace, based on the provisioner's returned endpoint info (performed by bucket lib).
+1. creation of a Secret, residing in the OBC's namespace, based on the provisioner's returned credentials (performed by bucket lib).
+
+### Custom Resource Details
+
+#### OBC Details (created by user)
+
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  name: MY-BUCKET-1 [1]
+  namespace: USER-NAMESPACE [2]
+spec:
+  bucketName: [3]
+  generateBucketName: "photo-booth" [4]
+  storageClassName: AN-OBJECT-STORE-STORAGE-CLASS [5]
+  additionalConfig: [6]
+    ANY_KEY: VALUE ...
+```
+1. name of the ObjectBucketClaim. This name becomes the name of the Secret and ConfigMap.
+1. namespace of the ObjectBucketClaim, which is also the namespace of the ConfigMap and Secret.
+1. name of the bucket. If supplied then `generateBucketName` is ignored.
+**Not** recommended for new buckets since names must be unique within
+an entire object store.
+1. if supplied then `bucketName` must be empty. This value becomes the prefix for a randomly generated name.
+After `Provision` returns `bucketName` is set to this random name.
+If both `bucketName` and `generateBucketName` are supplied then `BucketName` has precedence and `GenerateBucketName` is ignored. 
+If both `bucketName` and `generateBucketName` are blank or omitted then the storage class is expected to contain the name of an _existing_ bucket. It's an error if all three bucket related names are blank or omitted.
+1. storageClass which defines the object-store service and the bucket provisioner.
+1. additionalConfig gives providers a location to set store-specific config values (tenant, namespace...).
+The value is a list of 1 or more key-value pairs.
+
+**Note:** the OBC is mutated by the library with the addition of a finalizer and label:
+```
+  labels:
+    bucket-provisioner: PROVISIONER-NAME [1]
+  finalizers:
+  - "objectbucket.io/finalizer" [2]
+```
+1. the label value is the name of the provisioner but due to Kubernetes restrictions slash (/) is replaced by a dash (-).
+1. finalizers set and cleared by the lib's OBC controller. Prevents accidental deletion of an OB.
+
+#### OB Details (created by library)
+
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: ObjectBucket
+Metadata:
+  name: obc-my-ns-my-bucket [1]
+  namespace: namespace-of-object-store [2]
+  labels:
+    bucket-provisioner: PROVISIONER-NAME [3]
+  finalizers:
+  - "objectbucket.io/finalizer" [4]
+spec:
+  storageClassName: example-obj-prov [5]
+  claimRef: *v1.objectreference [6]
+  reclaimPolicy: {"Delete", "Retain"} [7]
+  endpoint:
+    bucketHost: foo.bar.com
+    bucketPort: 8080
+    bucketName: my-photos-1xj4a
+    region: # provisioner dependent
+    subRegion: # provisioner dependent
+    additionalConfigData: [] #string:string
```

Comment:
This list of endpoint parameters feels specific to a particular provider.  I feel like the goal of this should be to produce the fields that are needed by client SDKs and to be useful across at least the three main object APIs (S3, GCS, and Azure).   Given that, I'd argue that a "data path protocol" should be supplied with the claim (and declared by drivers).  The endpoint and credential information supplied by the driver (and recorded in the OB and Secret) should be appropriate for the given data path protocol (e.g., nearly directly usable by the standard API for that data protocol).

Reply:
+1 It sounds like you're describing a pattern similar to the *VolumeSource subfield of PVs, which I agree would be a better fit.  The overuse of string:string maps is something we need to address generally.

Reply:
@alarge:
> "data path protocol" should be supplied with the claim (and declared by drivers).

A goal was to not embed specific object store fields into the lib (though admittedly there is an s3 flavor). The reason for this is so object store vendor would not have to wait for their fields to be added. IOW, the lib would have some `switch dataPathProtocol` block which has to know which fields go with which protocol -- at least that's how I'm seeing. Can you elaborate more on this?